felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -1122,87 +1131,37 @@ func (m *bpfEndpointManager) onInterfaceUpdate(update *ifaceStateUpdate) {
 		return
 	}
 
+	// Should be safe without the lock since there shouldn't be any active background threads
+	// but taking it now makes us robust to refactoring.
+	m.ifacesLock.Lock()
+	defer m.ifacesLock.Unlock()
 	masterIfIndex := 0
-	prevMasterIfIndex := 0
 	curIfaceType := IfaceTypeUnknown
-	prevIfaceType := IfaceTypeUnknown
-	if val, ok := m.nameToIface[update.Name]; ok {
-		prevIfaceType = val.info.ifaceType
-		prevMasterIfIndex = val.info.masterIfIndex
-	}
-
-	if update.State != ifacemonitor.StateNotPresent && !m.isWorkloadIface(update.Name) {
-		// Determine the type of interface.
-		// These include host, bond, slave, ipip, wireguard, l3.
-		// update the ifaceType, master ifindex if bond slave.
-		link, err := m.dp.getIfaceLink(update.Name)
-		if err != nil {
-			log.Errorf("Failed to get interface information via netlink '%s'", update.Name)
-			curIfaceType = IfaceTypeL3
-			if m.isDataIface(update.Name) {
-				curIfaceType = IfaceTypeData
-			}
-		} else {
-			curIfaceType = m.getIfaceTypeFromLink(link)
-			masterIfIndex = link.Attrs().MasterIndex
-		}
-		if prevIfaceType != curIfaceType {
-			if curIfaceType == IfaceTypeBondSlave {
-				// Remove the Tc program.
-				ai, err := bpf.ListCalicoAttached()
-				if err != nil {
-					log.WithError(err).Warn("Failed to list attached programs")
-				} else {
-					if err := m.cleanupOldAttach(update.Name, ai[update.Name]); err != nil {
-						log.WithError(err).Warnf("Failed to detach old programs from now bonding device '%s'", update.Name)
-					}
-				}
-			} else if curIfaceType == IfaceTypeBond {
-				// create an entry in the hostIfaceToSlaveDevices
-				if _, ok := m.hostIfaceToSlaveDevices[update.Name]; !ok {
-					m.hostIfaceToSlaveDevices[update.Name] = set.New[string]()
-				}
-			}
-		}
-	}
-
-	// Manage bond slaves.
-	if !m.isWorkloadIface(update.Name) && curIfaceType != prevIfaceType {
-		if curIfaceType == IfaceTypeBondSlave {
-			/* Interface has been added to a bond.
-			 * Add this interface to the list of slave devices
-			 * of the master.
-			 */
-			netiface, err := m.dp.interfaceByIndex(masterIfIndex)
+	if !m.isWorkloadIface(update.Name) {
+		if update.State != ifacemonitor.StateNotPresent {
+			// Determine the type of interface.
+			// These include host, bond, slave, ipip, wireguard, l3.
+			// update the ifaceType, master ifindex if bond slave.
+			link, err := m.dp.getIfaceLink(update.Name)
 			if err != nil {
-				log.WithError(err).Warn("Failed to get master interface name. Slave devices not updated")
-			} else {
-				// Slave interface update comes first.
-				if _, ok := m.hostIfaceToSlaveDevices[netiface.Name]; !ok {
-					m.hostIfaceToSlaveDevices[netiface.Name] = set.New[string]()
+				log.Errorf("Failed to get interface information via netlink '%s'", update.Name)
+				curIfaceType = IfaceTypeL3
```

Comment:
why do we default to L3?

Reply:
when we don't get the link information, I default the type to L3 and if the interface matches in `dataIfacePattern`, I change it to Data.